airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py

```diff
@@ -0,0 +1,280 @@
+            message_repository=self.message_repository,  # type: ignore  # message_repository is always instantiated with a value by factory
+        )
+
+    def read(
+        self,
+        logger: logging.Logger,
+        config: Mapping[str, Any],
+        catalog: ConfiguredAirbyteCatalog,
+        state: Optional[Union[List[AirbyteStateMessage], MutableMapping[str, Any]]] = None,
+    ) -> Iterator[AirbyteMessage]:
+
+        # ConcurrentReadProcessor pops streams that are finished being read so before syncing, the names of the concurrent
+        # streams must be saved so that they can be removed from the catalog before starting synchronous streams
+        concurrent_stream_names = set([concurrent_stream.name for concurrent_stream in self._concurrent_streams])
+
+        selected_concurrent_streams = self._select_streams(streams=self._concurrent_streams, configured_catalog=catalog)
+        # It would appear that passing in an empty set of streams causes an infinite loop in ConcurrentReadProcessor.
+        # This is also evident in concurrent_source_adapter.py so I'll leave this out of scope to fix for now
+        if selected_concurrent_streams:
+            yield from self._concurrent_source.read(selected_concurrent_streams)
+
+        # Sync all streams that are not concurrent compatible. We filter out concurrent streams because the
+        # existing AbstractSource.read() implementation iterates over the catalog when syncing streams. Many
+        # of which were already synced using the Concurrent CDK
+        filtered_catalog = self._remove_concurrent_streams_from_catalog(catalog=catalog, concurrent_stream_names=concurrent_stream_names)
+        yield from super().read(logger, config, filtered_catalog, state)
+
+    def check(self, logger: logging.Logger, config: Mapping[str, Any]) -> AirbyteConnectionStatus:
+        return super().check(logger=logger, config=config)
+
+    def discover(self, logger: logging.Logger, config: Mapping[str, Any]) -> AirbyteCatalog:
+        return AirbyteCatalog(streams=[stream.as_airbyte_stream() for stream in self.all_streams(config=config)])
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        """
+        Returns the list of streams that can be run synchronously in the Python CDK.
+
+        NOTE: For ConcurrentDeclarativeSource, this method only returns synchronous streams because it usage is invoked within the
+        existing Python CDK. Streams that support concurrency are started from read().
+        """
+        return self._synchronous_streams
+
+    def all_streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        return self._synchronous_streams + self._concurrent_streams  # type: ignore  # Although AbstractStream doesn't inherit stream, they were designed to fit the same interface when called from streams()
+
+    def _separate_streams(self, config: Mapping[str, Any]) -> Tuple[List[AbstractStream], List[Stream]]:
```

Comment:
Do you think the name `_group_streams` would be more informative?

Reply:
i'm fine with `_group_streams`